agent/textwatcher.go

```diff
@@ -0,0 +1,194 @@
+package agent
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"hash/crc32"
+)
+
+type TextWatcher interface {
+	Start()
+	Stop()
+	Contents() ([]byte, error)
+	Path() string
+	Kind() string
+	MarshalJSON() ([]byte, error)
+}
+
+type textWatcher struct {
+	sync.Mutex
+	keepPolling  bool
+	kind         string
+	path         string
+	UpdatedAt    time.Time
+	BeingWatched bool
+	OnChange     ChangeHandler
+	Checksum     uint32
+	CmdName      string
+	CmdArgs      []string
+	contents     func() ([]byte, error)
+	pollSleep    time.Duration
+}
+
+// File watchers track changes in the contents of a file
+func NewFileWatcher(path string, callback ChangeHandler) Watcher {
+	var kind string
+	filename := filepath.Base(path)
+	switch filename {
+	case "Gemfile.lock":
+		kind = "gemfile"
+	case "available":
+		//todo support debian
+		kind = "ubuntu"
+	case "status":
+		kind = "ubuntu"
+	}
+
+	watcher := &textWatcher{
+		path:      path,
+		OnChange:  callback,
+		kind:      kind,
+		UpdatedAt: time.Now(),
+		pollSleep: env.PollSleep,
+	}
+	watcher.contents = watcher.FileContents
+
+	// Do a scan off the bat so we get a checksum, and PUT the file
+	watcher.scan()
+	return watcher
+}
+
+// Process watchers track changes in the output of a command
+func NewCommandOutputWatcher(process string, callback ChangeHandler) Watcher {
+	splat := strings.Split(process, " ")
+	name := splat[0]
+	args := splat[1:]
+
+	watcher := &textWatcher{
+		path:      process,
+		OnChange:  callback,
+		kind:      "centos",
```

Comment:
i.e. see above re: kind being useless. It'd make more sense to set kind to be the _kind of watcher_.

This will matter once we support a non-ruby, non-linux.